src/inference/gibbs.jl

```diff
@@ -41,20 +41,27 @@ Tips:
 methods like Particle Gibbs. You can increase the effectiveness of particle sampling by including
 more particles in the particle sampler.
 """
-struct Gibbs{space, A<:Tuple} <: InferenceAlgorithm
+struct Gibbs{space, A<:Tuple, B<:Tuple} <: InferenceAlgorithm
     algs::A   # component sampling algorithms
-
-    function Gibbs{space, A}(algs::A) where {space, A<:Tuple}
+    iterations::B
+    function Gibbs{space, A, B}(algs::A, iterations::B) where {space, A<:Tuple, B<:Tuple}
         all(isgibbscomponent, algs) || error("all algorithms have to support Gibbs sampling")
-        return new{space, A}(algs)
+        return new{space, A, B}(algs, iterations)
     end
 end
 
-function Gibbs(algs...)
+function Gibbs(args...)
+    if eltype(args) <: InferenceAlgorithm
+        algs = args
+        iterations = tuple(fill(1, length(args))...)
+    else
+        algs = tuple(map(first, args)...)
+        iterations = tuple(map(last, args)...)
+    end
```

Comment:
Maybe it would be a bit cleaner to use different constructors for both use cases with differently typed Varargs. I also wonder if one should actually support `::Tuple{InferenceAlgorithm, Int}` or if `Gibbs(::NTuple{N,InferenceAlgorithm}, ::NTuple{N, Int})` would be sufficient.

I also think a default of `nothing` would be better than `tuple(fill(1, ...))` (which should rather be `ntuple(_ -> 1, length(args))` to be inferrable I guess). Statements such as `if x.iterations !== nothing` could be compiled away completely if one just uses standard Gibbs, one could dispatch on it, and it's a simpler type.

Reply:
For interfaces, we have a few options:
1. `Gibbs(HMC(0.1, 2, :x), PG(30, :y), MH(:z), (10, 20, 30))
2. `Gibbs((HMC(0.1, 2, :x), 10), (PG(30, :y), 20), (MH(:z), 30)
3. `Gibbs(HMC(0.1, 2, :x) => 10, PG(30, :y) => 20, MH(:z), => 30)
The current implementation supports both 2 and 3.
I prefer it for 2 reasons:
1. It's similar to the old interface syntax: `Gibbs(HMC(0.1, 2, 10, :x), PG(30, 20, :y, MH(30, :z)`
2. The relation between each Gibbs component and the corresponding number of iterations is easy to read as they are next to each other. If we do Option 1, one need to do `zip` in their mind to see how many iterations each component needs to run.

I'd perfer to set iterations to be `1` if it's not specified, as it is what the Gibbs loop really does: iterates each component for 1 iteration, which is just a special case of what's added in this PR.

Reply:
> The relation between each Gibbs component and the corresponding number of iterations is easy to read in Option 2 and 3 as they are next to each other

I'm fine with that but I still think that it would be better to add specific constructors for the supported options with `Vararg`s of the desired types. I.e., one with `Gibbs(alg1::InferenceAlgorithm, algs::InferenceAlgorithm...)` and one with `Gibbs(alg1::Tuple{InferenceAlgorithm, Int}, algs::Tuple{InferenceAlgorithm, Int}...)` (the first algorithm is written separately since otherwise one (re-)defines `Gibbs()`). Or with `Pair{...}` - but also I don't think one should support multiple options but rather decide for one to keep the API clean.

> I'd perfer to set iterations to be 1 if it's not specified, as it is what the Gibbs loop really does

Sure, that's how the "standard" case can be represented in the more general formulation. But it's not clear that this is the best way to implement the default case. The main motivation for using `nothing` instead is that 1) it allows the compiler to optimize the generated code for the default case (without looping and additional type stability analysis and potential problems) and 2) allows developers to dispatch on the default vs general case (not sure when this would be helpful but it's at least an option). So I think that `nothing` would be a better choice.

Reply:
> I'm fine with that but I still think that it would be better to add specific constructors for the supported options with `Varargs` of the desired types.

I agree with you the type signatures shall be improved.
Let me do what you suggest.

> Sure, that's how the "standard" case can be represented in the more general formulation. But it's not clear that this is the best way to implement the default case.

I'm not sure how much performance gain this optimization would actually give. The main computation happens in each Gibbs component and I don't feel getting rid of the for loop actually does much.